### PR TITLE
Upgrade dynamic loading API

### PR DESCRIPTION
**Pull request by isaachier**
## Which problem is this PR solving?
- #118 

## Short description of the changes
- Update Hunter and upgrade dynamic loading API.

### CODE REVIEW

File: src/jaegertracing/DynamicLoad.cpp [hunk 1]
```diff
@@ -19,31 +19,37 @@
...
 #ifndef JAEGERTRACING_WITH_YAML_CPP
     *errorCategory =
         static_cast<const void*>(&opentracing::dynamic_load_error_category());
     return opentracing::dynamic_load_not_supported_error.value();
 #endif
-    if (std::strcmp(opentracingVersion, OPENTRACING_VERSION) != 0) {
-        *errorCategory = static_cast<const void*>(
+    if (std::strcmp(opentracing_version, OPENTRACING_VERSION) != 0 ||
+        std::strcmp(opentracing_abi_version, OPENTRACING_ABI_VERSION) != 0) {
+        *error_category = static_cast<const void*>(
```
Comment (rnburn) [inline src/jaegertracing/DynamicLoad.cpp]: I don't think you want `std::strcmp(opentracing_version, OPENTRACING_VERSION) != 0` in the comparison.

The point of adding the ABI version was to allow the plugin to work with other versions of OT if there's no ABI breakage.

The version is only passed in for the case that you might have some sort of incompatible semantic change in the future.

File: src/jaegertracing/DynamicLoad.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+#ifndef JAEGERTRACING_DYNAMIC_LOAD_H
+#define JAEGERTRACING_DYNAMIC_LOAD_H
+
+int JaegerTracingMakeTracerFactoryFct(const char* opentracing_version,
+                                      const char* opentracing_abi_version,
+                                      const void** error_category,
+                                      void* error_message,
+                                      void** tracer_factory);
```
Comment (rnburn) [inline src/jaegertracing/DynamicLoad.h]: You don't need the function exposes. The macro OPENTRACING_DECLARE_IMPL_FACTORY declares a variable function pointer (See for example how it's done in the LS tracer: https://github.com/lightstep/lightstep-tracer-cpp/blob/master/src/dynamic_load.cpp#L6)

File: src/jaegertracing/DynamicLoad.cpp (hunk 1 above)
Comment (isaachier) [reply src/jaegertracing/DynamicLoad.cpp]: Good point.

File: src/jaegertracing/DynamicLoad.h (hunk 2 above)
Comment (isaachier) [reply src/jaegertracing/DynamicLoad.h]: Thanks for the example!

**Comment by isaachier** [comment]
@rnburn do you mind checking if everything looks OK? Thanks in advance.

File: src/jaegertracing/DynamicLoad.cpp [hunk 3]
```diff
@@ -19,21 +19,23 @@
...
+                             const void** errorCategory,
+                             void* errorMessage,
+                             void** tracerFactory)
 {
     assert(errorCategory != nullptr);
     assert(tracerFactory != nullptr);
 #ifndef JAEGERTRACING_WITH_YAML_CPP
     *errorCategory =
-        static_cast<const void*>(&opentracing::dynamic_load_error_category());
+        static_cast<const void*>(&opentracing::dynamic_load_errorCategory());
```
Comment (rnburn) [inline src/jaegertracing/DynamicLoad.cpp]: dynamic_load_errorCategory?

Comment (isaachier) [reply src/jaegertracing/DynamicLoad.cpp]: Lol I was camelcasing this stuff an must have messed that up. Surprised this built...

**Pull request merged**